firecrown/parser.py

```diff
@@ -1,30 +1,48 @@
 import importlib
 import yaml
 import jinja2
+import collections
 
-
-def parse(filename):
+def parse(source, memory_data=None):
     """Parse a configuration file.
 
     Parameters
     ----------
-    filename : str
+    source : str
         The config file to parse. Should be YAML formatted.
+        if source has 'read' attribute, it is used to fetch data,
+        otherwise it is interpreted as a filename
+    memory_data : (optional). Dictionary (can be of dictionaries) 
+                  used to update parsed config file after reading it.
 
     Returns
     -------
-    config: dict
+    config : dict
         The raw config file as a dictionary.
     data : dict
         A dictionary containg each analyses key replaced with its
         corresponding data and function to compute the log-likelihood.
     """
 
-    with open(filename, 'r') as fp:
-        config_str = jinja2.Template(fp.read()).render()
+    if  not hasattr(source,"read"):
+        source = open (source,'r')
+    config_str = jinja2.Template(source.read()).render()
     config = yaml.load(config_str, Loader=yaml.Loader)
     data = yaml.load(config_str, Loader=yaml.Loader)
 
+    ## if stuff came already parsed, bring it in
+    if memory_data is not None:
+        ## recursive update of potentially nested dictionaries
+        def update_dict(d, u):
+            for k, v in u.items():
+                if isinstance(v, collections.abc.Mapping):
+                    d[k] = update_dict(d.get(k, {}), v)
+                else:
+                    d[k] = v
+            return d
+        data = update_dict (data, memory_data)
+        config = update_dict (config, memory_data)
```

Comment:
I need to think through this change and what it might mean. Letting people make essentially arbitrary updates to the config structure and the data could easily break things elsewhere in unexpected ways.